### PR TITLE
Add partial decoding function for improved streaming functionality

### PR DESCRIPTION
**Pull request by lopez86**
Currently, for streaming, we allow for the user to input a starting LM state. This might work ok in some cases but there are some issues:

We start from an empty beam every time - this is fine if there is not much correlation from chunk to chunk, but for streaming we ideally want a mode where the logits can be split up without affecting the results. So, we need an endpoint where the full beam information is included and the scoring caches are preserved between calls until the user clears them.

The code to use this will be a bit different since the user has to do more management of state objects. An alternate approach would be to save the state within the decoder object so that the user just has to call initialize and clear functions or using kwargs in the partial decode function - I'm not sure what's better here

This issue was pointed out a while ago in an issue but hasn't been addressed yet.

### CODE REVIEW

File: pyctcdecode/decoder.py [hunk 1]
```diff
@@ -609,6 +665,65 @@ def _decode_logits(
...
+        language_model = self._language_model
+        if language_model is None:
+            cached_lm_scores: LMScoreCache = {}
+        else:
+            start_state = language_model.get_start_state()
+            cached_lm_scores = {("", False): (0.0, 0.0, start_state)}
+        cached_p_lm_scores: Dict[str, float] = {}
+        return start_beam, cached_lm_scores, cached_p_lm_scores
+
+    def partial_decode_beams(
```
Comment (lopez86) [inline pyctcdecode/decoder.py:680]: this approach depends on the user for management of state like caches, previous beams, and the hotword sscorer. 

Another approach would be to add `initialize_partial_decoding_state()` and `clear_partial_decoding_state()` functions and preserving these things within the decoder object

A third approach would be to add an `initialize` kwarg to run on the first call and then clear state when `is_end==True`. 

Not sure which of these approaches would be best, they're all more or less equivalent but would be used differently by the user

File: pyctcdecode/decoder.py [hunk 2]
```diff
@@ -609,6 +665,65 @@ def _decode_logits(
...
+        self,
+        logits: NDArray[NpFloat],
+        cached_lm_scores: LMScoreCache,
+        cached_p_lm_scores: Dict[str, float],
+        beams: List[Beam],
+        beam_width: int = DEFAULT_BEAM_WIDTH,
+        beam_prune_logp: float = DEFAULT_PRUNE_LOGP,
+        token_min_logp: float = DEFAULT_MIN_TOKEN_LOGP,
+        prune_history: bool = DEFAULT_PRUNE_BEAMS,
+        hotword_scorer: Optional[HotwordScorer] = None,
```
Comment (lopez86) [inline pyctcdecode/decoder.py:690]: here I allow the user to directly input a hotword scorer object. in the case of partial decoding, a typical use case would be to repeatedly use the same hotwords between calls to this function for a single audio file, so we don't want to be building an identical hotword scorer over and over

File: pyctcdecode/decoder.py (hunk 1 above)
Comment (lopez86) [reply pyctcdecode/decoder.py:680]: after doing some more thinking, I think this approach is the correct one to go with - this allows the most freedom to the user. if state is held within the decoder, then a decoder object can only process one audio stream at a time. this way allows the decoder to process many streams - the state just has to be preserved somewhere else and called up later. this would also allow for better scaling. for a language model where the state is saved in a way that is safe for multiprocessing, we could run many decoder servers and it wouldn't matter which one a given chunk went to if the state is made accessible to all

File: pyctcdecode/decoder.py (hunk 2 above)
Comment (andrew-titus) [reply pyctcdecode/decoder.py:692]: This is neat, too, as one could also use different hotword scorers over the course of an audio file if they have some reasonable prior about word distribution

File: pyctcdecode/tests/test_decoder.py [hunk 3]
```diff
@@ -512,6 +512,51 @@ def test_beam_results(self):
         self.assertEqual(beams[0].text, "bugs bunny")
         self.assertEqual(beams[1].text, "bunny bunny")
 
+    def test_partial_decode(self):
```
Comment (andrew-titus) [inline pyctcdecode/tests/test_decoder.py:515]: Could we add a test or two for the hotword scorer stuff? Like maybe verify that we get the same results for a logits slice when providing the same hotword scorer, but different ones when providing different ones over the course of partial decoding?

**Review by andrew-titus — COMMENTED** [review]
Overall LGTM; just a comment and a test request

**Comment by lopez86** [comment]
Waiting on [this PR](https://github.com/kensho-technologies/pyctcdecode/pull/112) to fix CI and unblock this

**Comment by lopez86** [comment]
> Waiting on [this PR](https://github.com/kensho-technologies/pyctcdecode/pull/112) to fix CI and unblock this

Looks like it's ready to go now

**Review by andrew-titus — APPROVED** [review]
LGTM
